### PR TITLE
Express 4 support

### PR DESCRIPTION
**Pull request by devoidfury**
Updated for express 4 support. Dropped connect dependency.

Note: express 4.x dropped support for node v0.8.

Known issues:
* caching middleware or minify middleware seems to be junking binary files (fonts).

Please review and let me know if you see any issues.

Fix for issue #1738

### CODE REVIEW

**Comment by JohnMcLear** [comment]
Note to self when merging:  We support IO.JS and Node 12 too, I need to update files to show that..

**Comment by JohnMcLear** [comment]
@devoidfury if you can get the bits in about us supporting nodejs and iojs that would be useful then I can straight up merge :)

File: README.md [hunk 1]
```diff
@@ -22,7 +22,7 @@ Also, check out the **[FAQ](https://github.com/ether/etherpad-lite/wiki/FAQ)**,
 
 # Installation
 
-Etherpad works with node v0.8, v0.10 and v0.11, only. (We don't support v0.6)
+Etherpad works with node v0.10 and v0.11, only. (We don't support v0.6 or v0.8)
```
Comment (JohnMcLear) [inline README.md]: we do support 0.12

File: bin/installOnWindows.bat [hunk 2]
```diff
@@ -8,7 +8,7 @@ cmd /C node -e "" || ( echo "Please install node.js ( http://nodejs.org )" && ex
 
 echo _
 echo Checking node version...
-set check_version="if(['8','10'].indexOf(process.version.split('.')[1].toString()) === -1) { console.log('You are running a wrong version of Node. Etherpad requires v0.8.x or v0.10.x'); process.exit(1) }"
+set check_version="if(['10'].indexOf(process.version.split('.')[1].toString()) === -1) { console.log('You are running a wrong version of Node. Etherpad requires v0.10.x'); process.exit(1) }"
```
Comment (JohnMcLear) [inline bin/installOnWindows.bat]: afaik we require 0.10+ not 0.10

**Comment by devoidfury** [comment]
Updated readme and engine checking in the scripts for node v0.10+ and io.js.

**Comment by JohnMcLear** [comment]
Hitting up ``http://127.0.0.1:9001/tests/frontend/``

returns

``express deprecated res.sendfile: Use res.sendFile instead src/node/hooks/express/tests.js:53:9``

**Comment by JohnMcLear** [comment]
After replacing ``sendfile`` with ``sendFile`` the /tests/frontend page is still erroring with 

```
ReferenceError: jQuery is not defined sendkeys.js:349:0
ReferenceError: jQuery is not defined jquery.iframe.js:40:0
ReferenceError: $ is not defined
```

**Comment by JohnMcLear** [comment]
Trying to visit http://127.0.0.1:9001/p/test delivers a lot of 404's too
```
GET 
http://127.0.0.1:9001/static/css/pad.css [HTTP/1.1 404 Not Found 2ms]
GET 
http://127.0.0.1:9001/static/custom/pad.css [HTTP/1.1 404 Not Found 1ms]
GET 
http://127.0.0.1:9001/static/js/html10n.js [HTTP/1.1 404 Not Found 38ms]
GET 
http://127.0.0.1:9001/static/js/l10n.js [HTTP/1.1 404 Not Found 41ms]
GET 
http://127.0.0.1:9001/static/js/require-kernel.js [HTTP/1.1 404 Not Found 1ms]
GET 
http://127.0.0.1:9001/static/custom/pad.js [HTTP/1.1 404 Not Found 39ms]
GET 
http://127.0.0.1:9001/static/js/l10n.js [HTTP/1.1 404 Not Found 2ms]
GET 
http://127.0.0.1:9001/static/js/require-kernel.js [HTTP/1.1 404 Not Found 1ms]
Password fields present on an insecure (http://) page. This is a security risk that allows user login credentials to be stolen.[Learn More] <unknown>
ReferenceError: require is not defined pad.js:1:0
ReferenceError: require is not defined ace2_common.js:1:0
GET 
http://127.0.0.1:9001/static/custom/pad.js [HTTP/1.1 404 Not Found 1ms]
ReferenceError: require is not defined test:531:14
```

File: src/node/hooks/express/static.js [hunk 3]
```diff
@@ -9,11 +9,11 @@ exports.expressCreateServer = function (hook_name, args, cb) {
 
   // Cache both minified and static.
   var assetCache = new CachingMiddleware;
-  args.app.all('/(javascripts|static)/*', assetCache.handle);
+  args.app.all(/\/(javascripts|static)\/(.*)/, assetCache.handle);
 
   // Minify will serve static files compressed (minify enabled). It also has
   // file-specific hacks for ace/require-kernel/etc.
-  args.app.all('/static/:filename(*)', minify.minify);
+  args.app.all('/static/:filename', minify.minify);
```
Comment (JohnMcLear) [inline src/node/hooks/express/static.js]: Missing ``/(.*)/`` ?

Comment (devoidfury) [inline src/node/hooks/express/static.js]: `'/static/:filename'` translates under the hood to `/^\/static\/(?:([^\/]+?))\/?$/i`, should this have slashes in it? If so, it should be `'/static/:filename+'`

**Comment by devoidfury** [comment]
Ah, ok. I didn't test 0.12, didn't see these issues in 0.10

File: src/node/utils/Minify.js [hunk 4]
```diff
@@ -145,7 +145,6 @@ function minify(req, res, next)
   filename = path.normalize(path.join(ROOT_DIR, filename));
   if (filename.indexOf(ROOT_DIR) == 0) {
     filename = filename.slice(ROOT_DIR.length);
-    filename = filename.replace(/\\/g, '/'); // Windows (safe generally?)
```
Comment (devoidfury) [inline src/node/utils/Minify.js]: Node fixes these paths automatically on windows, and it added blind spot

**Comment by JohnMcLear** [comment]
Looks good thanks! :) Testing now!

**Comment by JohnMcLear** [comment]
Found another one liner

import test is failing
```
importexport.js

    import functionality
        import a pad with newlines from txt136ms
        "before each" hook

        the string "Error: expected '' to equal '<span class="">imported text</span>\n<span class="">newline</span>\n<br>\n'" was thrown, throw an Error :)
        Runner.prototype.fail@http://127.0.0.1:9001/tests/frontend/lib/mocha.js:3786:1
        Runner.prototype.uncaught@http://127.0.0.1:9001/tests/frontend/lib/mocha.js:4059:3
        uncaught@http://127.0.0.1:9001/tests/frontend/lib/mocha.js:4089:5
```
Caused by:

``express deprecated res.send(body, status): Use res.status(status).send(body) instead src/node/hooks/express/webaccess.js:101:14``

**Comment by devoidfury** [comment]
I'll get that warning fixed, but that test failed on develop for me as well, on node v0.12 -- could you verify that?

**Comment by JohnMcLear** [comment]
Yea I will do, if it failed it's probably related to ``abiword`` being enabled or so..  I will check in a moment

**Comment by devoidfury** [comment]
There is a security issue, should I contact you in private or would you prefer to discuss it here?

**Comment by JohnMcLear** [comment]
please email security@etherpad.org and cc john@mclear.co.uk

Thanks!

**Comment by JohnMcLear** [comment]
It does look like the tests need some polish, just reviewing those now..

**Comment by JohnMcLear** [comment]
Excellent, thanks man!

**Comment by JohnMcLear** [comment]
Interestingly this broke the font file we use for icons..

**Comment by JohnMcLear** [comment]
Possibly caused by: http://stackoverflow.com/questions/3594823/mime-type-for-woff-fonts

**Comment by JohnMcLear** [comment]
@devoidfury any ideas on why the font files might be delivered incorrectly?  I tried setting the type on them but they show up as unreadable/corrupt in network?

**Comment by JohnMcLear** [comment]
Installing the plugin ep_tasklist also shows this error as the .png icon included in the plugin is unable to load?  Mime type is correct...  Response is null ;/

**Comment by devoidfury** [comment]
I think it's due to a change in the semantics of how binary files are handled in express and the local files, caching_middleware or Minify, where we hijack the request/response and juggle the file around.

**Comment by devoidfury** [comment]
I'm investigating it over here, too.

**Comment by devoidfury** [comment]
Ok, found it.

Looks like the old route path that the assetCache was hooked up to was misleading.

    args.app.all('/(javascripts|static)/*', assetCache.handle);

That actually didn't match paths like `/static/...`, but it did match `/javascripts/...`. I'm not entirely sure why, but I checked it by logging paths at the top of the handle function.

Changing the route to this seemed to fix it:

    args.app.all(/\/javascripts\/(.*)/, assetCache.handle);


This means the binary junking issue may have been present before the express migration, but because the old routepath only matched `/javascript/` it was never detected. I'll double-check that.

**Comment by JohnMcLear** [comment]
Awesome, wanna chuck through a new PR with a fix or should I patch develop directly?

**Pull request merged**